File: drills/elixir/solutions/nc40_min_interval__offline_by_length.exs
defmodule Solution do
  # Answer each query once, and never revisit it. Taking the intervals shortest
  # first means the first interval to cover a query is already its answer, so
  # every query leaves the pool the moment it is settled and the pool only ever
  # shrinks.
  def min_interval(intervals, queries) do
    numbered = Enum.with_index(queries)

    {answered, unanswered} =
      intervals
      |> Enum.sort_by(fn {s, e} -> e - s end)
      |> Enum.reduce({%{}, numbered}, fn {s, e}, {answers, waiting} ->
        {covered, still_waiting} =
          Enum.split_with(waiting, fn {query, _index} -> s <= query and query <= e end)

        answers =
          Enum.reduce(covered, answers, fn {_query, index}, acc ->
            Map.put(acc, index, e - s + 1)
          end)

        {answers, still_waiting}
      end)

    answers =
      Enum.reduce(unanswered, answered, fn {_query, index}, acc ->
        Map.put(acc, index, -1)
      end)

    Enum.map(0..(length(queries) - 1)//1, &Map.fetch!(answers, &1))
  end
end

File: drills/go/solutions/nc40_min_interval__offline_by_length.go
package main

import "sort"

func minInterval(intervals [][]int, queries []int) []int {
	// Take the intervals shortest first; the first one to cover a query is
	// its answer, so each query is settled exactly once. Points are found
	// by binary search over the sorted distinct queries.
	sorted := append([][]int{}, intervals...)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i][1]-sorted[i][0] < sorted[j][1]-sorted[j][0]
	})
	distinct := sortInts(queries)
	answer := map[int]int{}
	for _, interval := range sorted {
		from := sort.SearchInts(distinct, interval[0])
		for i := from; i < len(distinct) && distinct[i] <= interval[1]; i++ {
			if _, done := answer[distinct[i]]; !done {
				answer[distinct[i]] = interval[1] - interval[0] + 1
			}
		}
	}
	result := make([]int, len(queries))
	for i, q := range queries {
		if size, ok := answer[q]; ok {
			result[i] = size
		} else {
			result[i] = -1
		}
	}
	return result
}
